requests/utils.py

```diff
@@ -642,25 +642,26 @@ def should_bypass_proxies(url, no_proxy):
     netloc = urlparse(url).netloc
 
     if no_proxy:
-        # We need to check whether we match here. We need to see if we match
-        # the end of the netloc, both with and without the port.
         no_proxy = (
             host for host in no_proxy.replace(' ', '').split(',') if host
         )
 
-        ip = netloc.split(':')[0]
-        if is_ipv4_address(ip):
+        # We need to see if we match the end of the netloc, accounting for a
+        # 'username[:password]@' at the beginning and a ':port' at the end.
+
+        host_or_ip = netloc.split('@')[-1].split(':')[0]
```

Comment:
People are rather inconsistent as to whether or not they properly encode their password. If they have an `@` in their password, and they've not encoded it appropriately, this will cause issues. It would be better to `rsplit` here as a result. It'll cause less teeth-gnashing.